src/v/cluster/cloud_metadata/offsets_recoverer.cc

```diff
@@ -0,0 +1,231 @@
+      _connection_cache.local());
+    for (const auto& snapshot_path : snapshot_paths) {
+        auto offsets_snapshot = cloud_storage::remote_file(
+          _remote.local(),
+          _cache.local(),
+          req.bucket,
+          cloud_storage::remote_segment_path{snapshot_path},
+          retry_node,
+          "offsets_snapshot");
+        group_offsets_snapshot snapshot;
+        try {
+            auto f = co_await offsets_snapshot.hydrate_readable_file();
+            auto f_size = co_await f.size();
+            ss::file_input_stream_options options;
+            auto input = ss::make_file_input_stream(f, options);
+            auto snap_buf_parser = iobuf_parser{
+              co_await read_iobuf_exactly(input, f_size)};
+            snapshot = serde::read<group_offsets_snapshot>(snap_buf_parser);
+        } catch (...) {
+            reply.ec = cluster::errc::allocation_error;
+            co_return reply;
+        };
+        // Collect the NTPs corresponding to the snapshot file, in preparation
+        // to perform some offset lookups.
+        absl::btree_set<model::ntp> ntps;
+        for (const auto& g : snapshot.groups) {
+            for (const auto& tp : g.offsets) {
+                for (const auto& po : tp.partitions) {
+                    ntps.emplace(
+                      model::kafka_namespace, tp.topic, po.partition);
+                }
+            }
+        }
+        try {
+            co_await offsets_lookup.run_lookups(std::move(ntps), retry_node);
+        } catch (ss::timed_out_error&) {
+            reply.ec = cluster::errc::timeout;
+            co_return reply;
+        }
+        const auto& offsets_by_ntp = offsets_lookup.offsets_by_ntp();
+        group_offsets_snapshot batched_groups;
+        for (auto& g : snapshot.groups) {
+            // Collect into a batched request so we can periodically kick off a
+            // batch of commits.
+            batched_groups.groups.emplace_back(group_offsets{});
+            auto& req_group = batched_groups.groups.back();
+            req_group.group_id = g.group_id;
+
+            bool discard_group = false;
+            absl::flat_hash_map<model::topic, group_offsets::topic_partitions>
+              req_topics;
+            for (auto& tp : g.offsets) {
+                for (auto& po : tp.partitions) {
+                    auto ntp = model::ntp{
+                      model::kafka_namespace, tp.topic, po.partition};
+                    if (auto offset_it = offsets_by_ntp.find(ntp);
+                        offset_it != offsets_by_ntp.end()) {
+                        auto hwm = offset_it->second;
+                        if (po.offset > hwm) {
+                            // The recovered partition doesn't have all the data
+                            // required to restore the snapshot exactly. Trim
+                            // it down to the high watermark.
+                            po.offset = kafka::offset{hwm()};
```

Comment:
should there be a log here, or is it evident from logs in other places that we are trimming?

Reply:
Done